### PR TITLE
[1/N] Add -Wdeprecated and related fixes

### PR DESCRIPTION
**Pull request by cyyever**
This PR adds -Wdeprecated to CMake warnings and fixes related issues.
cc @EikanWang @jgong5 @ezyang @colesbury @Skylion007

### CODE REVIEW

File: aten/src/ATen/core/TensorBase.h [hunk 1]
```diff
@@ -205,18 +205,8 @@ class TORCH_API TensorBase {
...
-  TensorBase& operator=(const TensorBase& x) & {
-    impl_ = x.impl_;
-    return *this;
-  };
-  TensorBase& operator=(TensorBase&& x) & noexcept {
-    impl_ = std::move(x.impl_);
-    return *this;
-  }
-
-  // Ban assignment to rvalues, since at::Tensor (weirdly) performs a deep copy here
```
Comment (Skylion007) [inline aten/src/ATen/core/TensorBase.h:217]: Are we sure this is no longer an issue? Do we have a test case for this?

Comment (ezyang) [reply aten/src/ATen/core/TensorBase.h:217]: This is still legit. It won't be covered by tests because it is disabling a footgun

Comment (cyyever) [reply aten/src/ATen/core/TensorBase.h:217]: Assignment to rvalue is actually strange. This should be an indication of logic errors.

Comment (ezyang) [reply aten/src/ATen/core/TensorBase.h]: In Tensor, it's to handle idioms like this: `tensor[0] = blah`. The `tensor[0]` is an rvalue and the assignment lets us copy the data from blah into the slice. When @peterbell10 refactored TensorBase out, I guess he thought this behavior was weird and disabled it on TensorBase.

**Comment by ezyang** [comment]
@cyyever I noticed some of your other refactors are getting bounced off of main, and this one is starting to conflict. You might consider splitting this up into smaller fixes to land before flipping the switch on Wdeprecated. If you want, you can use ghstack to make coordinating them easier (though since the splits don't depend on each other, I personally prefer not to use ghstack in that case.)

File: c10/util/MaybeOwned.h [hunk 2]
```diff
@@ -157,7 +160,6 @@ class MaybeOwned final {
         isBorrowed_ = false;
       }
     }
-    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(isBorrowed_ == rhs.isBorrowed_);
```
Comment (ezyang) [inline c10/util/MaybeOwned.h:160]: What's going on here?

File: c10/util/MaybeOwned.h [hunk 3]
```diff
@@ -136,7 +136,10 @@ class MaybeOwned final {
   }
 
   MaybeOwned& operator=(MaybeOwned&& rhs) noexcept(
-      std::is_nothrow_move_assignable<T>::value) {
+      std::is_nothrow_move_assignable_v<T>&& std::is_nothrow_move_assignable_v<
+          borrow_type>&& std::is_nothrow_move_constructible_v<T>&&
+          std::is_nothrow_destructible_v<T>&&
+              std::is_nothrow_destructible_v<borrow_type>) {
```
Comment (ezyang) [inline c10/util/MaybeOwned.h:142]: @swolchok mind checking this?

**Review by ezyang — COMMENTED** [review]
This is too big. Split it up. I reviewed until MaybeOwned.h and ran out of steam.

File: c10/util/MaybeOwned.h (hunk 2 above)
Comment (cyyever) [reply c10/util/MaybeOwned.h:160]: In a debug build, it may throw,  breaking the noexcept specification

**Comment by cyyever** [comment]
@ezyang It was reduced to contain c10 changes.

File: c10/util/Optional.h [hunk 4]
```diff
@@ -42,20 +42,6 @@
...
-#endif
-#if C10_CLANG_HAS_WARNING("-Wshorten-64-to-32")
-C10_CLANG_DIAGNOSTIC_IGNORE("-Wshorten-64-to-32")
-#endif
-#if C10_CLANG_HAS_WARNING("-Wimplicit-float-conversion")
-C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-float-conversion")
-#endif
-#if C10_CLANG_HAS_WARNING("-Wimplicit-int-conversion")
-C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-int-conversion")
-#endif
```
Comment (ezyang) [inline c10/util/Optional.h:57]: No action needed, but are you working on moving us to std::optional too?

**Review by ezyang — APPROVED** [review]

File: c10/util/Optional.h (hunk 4 above)
Comment (cyyever) [reply c10/util/Optional.h:57]: @swolchok has a work in progress. I am waiting that to finish.

**Review by Skylion007 — APPROVED** [review]

File: c10/util/Optional.h (hunk 4 above)
Comment (cyyever) [reply c10/util/Optional.h:57]: @ezyang But I have a work moving to std::variant #103675

**Comment by clee2000** [comment]
@pytorchbot revert -m "I'm getting errors internally that look like the below on x86_64-apple-ios-simulator with clang 16" -c ghfirst
```
xplat/caffe2/c10/util/Optional.h:740:12: error: implicit conversion turns string literal into bool: 'const char[14]' to 'bool' [-Werror,-Wstring-conversion]
    return TR2_OPTIONAL_ASSERTED_EXPRESSION(initialized(), dataptr());
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                 ^
```

**Comment by cyyever** [comment]
@clee2000 Should we put x86_64-apple-ios-simulator tests in CI?

**Comment by clee2000** [comment]
I'm not sure what the string literal is tbh

Ed says 
```
TR2_OPTIONAL_HOST_CONSTEXPR T const& operator*() const& {
    return TR2_OPTIONAL_ASSERTED_EXPRESSION(initialized(), contained_val());
  }
```
is the critical line.

This is what the rest of the error looks like, I've changed some things since I'm not sure what I'm actually allowed to post 
```
In file included from ***
In file included from caffe2/caffe2/core/operator.h:21:
In file included from caffe2/caffe2/core/blob.h:14:
In file included from caffe2/caffe2/core/tensor.h:5:
In file included from caffe2/caffe2/core/storage.h:23:
In file included from caffe2/c10/core/Storage.h:3:
In file included from caffe2/c10/core/StorageImpl.h:4:
In file included from caffe2/c10/core/SymInt.h:3:
In file included from caffe2/c10/core/SymBool.h:3:
In file included from caffe2/c10/core/SymNodeImpl.h:6:
caffe2/c10/util/Optional.h:749:12: error: implicit conversion turns string literal into bool: 'const char[14]' to 'bool' [-Werror,-Wstring-conversion]
    return TR2_OPTIONAL_ASSERTED_EXPRESSION(initialized(), contained_val());
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
caffe2/c10/util/Optional.h:87:37: note: expanded from macro 'TR2_OPTIONAL_ASSERTED_EXPRESSION'
  ((CHECK) ? (EXPR) : ([] { assert(!#CHECK); }(), (EXPR)))
                                   ~^~~~~~
<scratch space>:85:1: note: expanded from here
"initialized()"
^~~~~~~~~~~~~~~
***/third-party/toolchains/xcode/14.1.0/MacOSX.sdk/usr/include/assert.h:99:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __ASSERT_FILE_NAME, __LINE__, #e) : (void)0)
                        ^
caffe2/c10/util/Optional.h:782:20: note: in instantiation of member function 'c10::optional<bool>::operator*' requested here
    return *this ? **this : detail_::convert<T>(constexpr_forward<V>(v));
                   ^
caffe2/c10/core/TensorOptions.h:42:24: note: in instantiation of function template specialization 'c10::optional<bool>::value_or<bool>' requested here
  return pinned_memory.value_or(false);
                       ^
6 errors generated.
    When running <c++ preprocess_and_compile>.
    When building rule ***
Not all rules succeeded.
```

I thought the ios simulator tests would have been enough but I guess not? @malfet 

cc @ezyang

**Comment by malfet** [comment]
@clee2000 our iOS build are running using different version of xcode, this is why it was not caught

**Pull request closed without merging**